targets/csharp/source/source/PlayFabGameserverInternal.cs

```diff
@@ -0,0 +1,324 @@
+            }
+
+            finalConfig[PlayFabGameserverSDK.HeartbeatEndpointKey] = localConfig.HeartbeatEndpoint;
+            finalConfig[PlayFabGameserverSDK.ServerIdKey] = localConfig.ServerId;
+            finalConfig[PlayFabGameserverSDK.LogFolderKey] = localConfig.LogFolder;
+            finalConfig[PlayFabGameserverSDK.SharedContentFolderKey] = localConfig.SharedContentFolder;
+            finalConfig[PlayFabGameserverSDK.CertificateFolderKey] = localConfig.CertificateFolder;
+            finalConfig[PlayFabGameserverSDK.TitleIdKey] = localConfig.TitleId;
+            finalConfig[PlayFabGameserverSDK.BuildIdKey] = localConfig.BuildId;
+            finalConfig[PlayFabGameserverSDK.RegionKey] = localConfig.Region;
+
+            return finalConfig;
+        }
+
+        private async Task HeartbeatAsync()
+        {
+            while (_heartbeatRunning)
+            {
+                if (_signalHeartbeatEvent.WaitOne(1000))
+                {
+                    if (_debug)
+                    {
+                        Logger.Log("State transition signaled an early heartbeat.");
+                    }
+
+                    _signalHeartbeatEvent.Reset();
+                }
+
+                await SendHeartbeatAsync();
+            }
+
+            Logger.Log("Shutting down heartbeat thread");
+            _heartbeatDoneEvent.Set();
+        }
+
+        internal async virtual Task SendHeartbeatAsync()
+        {
+            bool gameHealth = false;
+            if (HealthCallback != null)
+            {
+                gameHealth = HealthCallback();
+            }
+
+            var payload = new HeartbeatRequest
+            {
+                CurrentGameState = this.State,
+                CurrentGameHealth = gameHealth ? "Healthy" : "Unhealthy",
+                CurrentPlayers = ConnectedPlayers.ToArray(),
+            };
+
+            try
+            {
+                HeartbeatResponse response = await _webClient.SendHeartbeatAsync(payload);
+                await UpdateStateFromHeartbeatAsync(response);
+
+                if (_debug)
+                {
+                    Logger.Log($"Heartbeat request: {{ state = {payload.CurrentGameState} }} response: {{ operation = {response.Operation} }}");
+                }
+            }
+            catch (Exception ex)
+            {
+                Logger.Log($"Cannot send heartbeat: {ex.Message}\r\n\r\n{ex}");
+                return;
+            }
+        }
+
+        private Task UpdateStateFromHeartbeatAsync(HeartbeatResponse response)
+        {
+            if (response.SessionConfig != null)
+            {
+                ConfigMap.AddIfNotNullOrEmpty("sessionCookie", response.SessionConfig?.SessionCookie);
+                ConfigMap.AddIfNotNullOrEmpty("sessionId", response.SessionConfig?.SessionId.ToString());
+
+                if (response.SessionConfig?.InitialPlayers != null && response.SessionConfig.InitialPlayers.Any())
```

Comment:
>InitialPlayers [](start = 44, length = 14)

can we add a comment about why we do the check and not set it blindly? I remember I had asked the same question in a previous PR